### ISSUE TITLE
v2: interpretation of files outside of the spec

### ISSUE THREAD

**Issue by joshmoore** [zarr-developers/zarr-specs#112]
see: https://github.com/ome/ngff/issues/27

The Zarr specification currently only defines 4 leaf nodes: .zarray, .zgroups, .zattrs, and chunks.

One OME-Zarr client ([bioformats2raw](https://github.com/glencoesoftware/bioformats2raw)) currently places an XML file within the Zarr hierarchy. This can be accessed at the Store level via:
```
text = store["METADATA.xml"].decode() # manually decode ascii
```
but not from the given group. Questions:

* Was there previously any discussion about possible Zarr-internal interpretations of files?
* Does anyone else have similar use cases?
* If this is generally valuable, would it break anyone _else's_ use cases to define such files as "1-dimensional non-chunked arrays"

cc: @manzt

**Comment by rabernat** [comment]
Just noting the the newly announced ["nczarr" spec](https://www.unidata.ucar.edu/software/netcdf/documentation/NUG/nczarr_head.html) also uses additional files in the zarr hierarchy, so technically it is not compliant with the Zarr spec. cc @DennisHeimbigner

**Comment by joshmoore** [comment]
Thanks for the reminder, @DennisHeimbigner. I'll add that as an additional option.

**Comment by joshmoore** [comment]
Managed to discuss this issue with @alimanfoo today. The original intent in the spec was to be hacker friendly and not prohibit storing other keys _as long as_ they don't negatively influence the interpretation of existing objects.

On the question of round-tripping to HDF5, we would need to suggest a specification/protocol _on the HDF5 side_ to make this feasible.

Looking forward to comments, but as this starts to crystallize, I'll open a PR to suggest v2.rst language (if no one beats me to it).

cc: @manzt @JacksonMaxfield @chris-allan

**Comment by DennisHeimbigner** [comment]
> On the question of round-tripping to HDF5, we would need to suggest a specification/protocol on the HDF5 side to make this feasible.

Can someone clarify what is meant by the HDF5 round-tripping? Is there
an issue describing this?

**Comment by joshmoore** [comment]
@DennisHeimbigner: No, _mea culpa_. If you convert a Zarr file to HDF5, zarrays, zgroups and attributes all have a clear mapping. As far as I can tell, these additional files do _not_. One could certainly map them to a dataset on the HDF5 side, but then when mapping _back to_ Zarr, without additional metadata, they will become zarrays rather than additional keys.

**Comment by DennisHeimbigner** [comment]
My speculation is that translating HDF5 will run into the same extra metadata issues I have been addressing in NCZarr/netcdf-c. As as I can tell, the two options are to add extra metadata as new objects or to add new keys to existing objects. Or a combination.

If the HDF5 round-trip issue reaches critical mass, then perhaps we should consider a  conversation on what approach to adding extra metadata is best with respect to other implementations of Zarr.

**Comment by joshmoore** [comment]
Correct me if I'm wrong, @DennisHeimbigner, but it sounds like the issue your pointing out is that HDF5 is a superset of Zarr, right? :+1: for having a conversation about those mappings so that the round-tripping in the _other direction_ isn't lossy. As far as I know though, this is the only case of Zarr allowing something that's not possible to describe in HDF5.

**Comment by shoyer** [comment]
`.zmetadata` for consolidated metadata (supported by Zarr-Python) is another good example of such extra files: https://github.com/zarr-developers/zarr-python/issues/720

**Comment by joshmoore** [comment]
@shoyer, yup, but that's certainly one I could see adding to `v2.rst` to explain the format in case other implementations want to add support.

**Comment by adair-kovac** [comment]
I just came across this thread and wanted to chime in that we've been taking advantage of the fact that it's fine to store other file types in a zarr group. One use case is adding root groups on an ad-hoc basis––it's nice that if you have a bunch of zarr groups that were written separately to a shared directory, all you have to do to combine them into one store is to zarr.open the parent directory. 

E.g. in one case we had some data on S3 like this:

```
root -> subdir1 -> hdf5
     -> subdir2 (.zgroup) -> various zarr groups
```
       
We converted the hdf5 data in subdir1 to zarr, left the hdf5 for backwards compatibility, and put a .zgroup file into the root directory. So now the whole S3 bucket is a zarr store without affecting users of other data types that were already there.

```
root (.zgroup) -> subdir1 (.zgroup) -> hdf5
                                    -> converted data (.zgroup) -> zarr arrays
               -> subdir2 (.zgroup) -> various zarr groups                    
```

**Comment by joshmoore** [comment]
Thanks for the use case, @adair-kovac!

**Comment by joshmoore** [comment]
Noted, @TomNicholas. For V2, the ship has essentially sailed, I think, is the fair take-home from the above.

For V3, though, while working on ZEP10, I was considering a "file-extension" as an example, basically a mechanism to register in the metadata, "I'm going to create a file at this path and here's what, if anything, you need to know about it".

**Comment by lubianat** [comment]
Is there any update of the interpretation here? We were discussing thumbnails in the context of (OME-)Zarr here: https://github.com/clbarnes/zarr-convention-thumbnails (@clbarnes) and we assumed it was not strictly forbidden. Is it? 

This is probably also relevant for other OME-Zarr discussions, as some tooling rely on an [OME-XML file](https://ngff.openmicroscopy.org/specifications/0.5/index.html#bioformats2raw-layout-transitional).

**Comment by jbms** [comment]
In general I would say that zarr v2 is unlikely to see any significant official changes and therefore additional files are unlikely to create any conflict with new official versions of zarr v2.  With zarr v3 there is some risk of conflict with future versions of the spec.
